### PR TITLE
Update Version regex to support ranges

### LINKED ISSUE #345: Version Ranges: Update CRD Regex
### Summary
The regular expression contained in the CRD that clients use to validate input before sending to the server must be updated to support version ranges. Due to the complexity and imprecision of the hyphen (-) operator, it will not be included.

* Update the regex to support the *Masterminds/semver* syntax (except for hyphen ranges).
* Add unit tests to validate ranges, both positive and negative. This must use a client program (e.g. `oc` or `kubectl`).

### Reference

[RFC: OLMv1 Version Ranges](https://docs.google.com/document/d/1Cf4vw6kdadNp3bHFxh5AMcq1xq0vsEUgC-JNEaVtU6g/edit?usp=sharing)

Update the Operator CR to support version ranges. There are no direct API changes.

### Related GitHub Issues

See dependencies.

### Dependencies

#344 
#346 may remove the regex

### Updates and Amendments

None.

### PR DESCRIPTION
**Pull request by tmshort**
Fixes #345

Add positive and negative test cases.

# Description

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

### CODE REVIEW

File: internal/controllers/admission_test.go [hunk 1]
```diff
@@ -61,6 +61,14 @@ var _ = Describe("Operator Spec Validations", func() {
...
 			"1.2.-3",
 			".1.2.3",
+			"<<1.2.3",
+			">>1.2.3",
+			">~1.2.3",
+			"==1.2.3",
+			"=!1.2.3",
+			"1.Y",
+			">1.2.3 && <2.3.4",
+			">1.2.3;<2.3.4",
```
Comment (joelanford) [inline internal/controllers/admission_test.go:71]: Should we add some valid masterminds range constraints to _our_ invalid list? (e.g. `"1.2.3-2.0.0"`)

**Review by joelanford — APPROVED** [review]
A few nits. But lgtm!

File: internal/controllers/admission_test.go (hunk 1 above)
Comment (tmshort) [reply internal/controllers/admission_test.go:71]: I was _just_ thinking about the hyphen range!
So, probably yes.

File: internal/controllers/admission_test.go [hunk 2]
```diff
@@ -69,7 +77,51 @@ var _ = Describe("Operator Spec Validations", func() {
 			}))
 
 			Expect(err).To(HaveOccurred(), "expected error for invalid semver %q", invalidSemver)
-			Expect(err.Error()).To(ContainSubstring("spec.version in body should match '^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(-(0|[1-9]\\d*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(\\.(0|[1-9]\\d*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*)?(\\+([0-9a-zA-Z-]+(\\.[0-9a-zA-Z-]+)*))?$'"))
+			Expect(err.Error()).To(ContainSubstring("spec.version in body should match '^(\\s*(=||!=|>|<|>=|=>|<=|=<|~|~>|\\^)\\s*(v?(0|[1-9]\\d*|[x|X|\\*])(\\.(0|[1-9]\\d*|x|X|\\*]))?(\\.(0|[1-9]\\d*|x|X|\\*))?(-([0-9A-Za-z\\-]+(\\.[0-9A-Za-z\\-]+)*))?(\\+([0-9A-Za-z\\-]+(\\.[0-9A-Za-z\\-]+)*))?)\\s*)((?:\\s+|,\\s*|\\s*\\|\\|\\s*)(=||!=|>|<|>=|=>|<=|=<|~|~>|\\^)\\s*(v?(0|[1-9]\\d*|x|X|\\*])(\\.(0|[1-9]\\d*|x|X|\\*))?(\\.(0|[1-9]\\d*|x|X|\\*]))?(-([0-9A-Za-z\\-]+(\\.[0-9A-Za-z\\-]+)*))?(\\+([0-9A-Za-z\\-]+(\\.[0-9A-Za-z\\-]+)*))?)\\s*)*$'"))
```
[the remark this answers is not in the document]
Comment (tmshort) [reply internal/controllers/admission_test.go:80]: Probably not, there's only one important regex per fiield; it's the field name that's important.

**Comment by tmshort** [comment]
@joelanford made requested changes.

**Comment by joelanford** [comment]
LGTM

However, do we need to keep this PR held open until the backing implementation actually uses masterminds? I'm guessing if we merge without masterminds in the picture, the result is that masterminds-but-not-blang ranges will result in a reconcile error instead of a validation error?

**Comment by tmshort** [comment]
> However, do we need to keep this PR held open until the backing implementation described in #346 is implemented? I'm guessing if we merge without masterminds in the picture, the result is that masterminds-but-not-blang ranges will result in a reconcile error instead of a validation error?

Yes, but it's necessary to be able to test ("get to") the reconcile code.

**Comment by tmshort** [comment]
@joelanford since there's a PR up for the MM code changes (updates), do you have any objections to this being merged?

**Comment by tmshort** [comment]
Closing in favor of #374 (which includes this).
